src/mem/ruby/structures/ALUFreeListArray.cc

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (c) 2023 The University of Wisconsin
+ *
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met: redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer;
+ * redistributions in binary form must reproduce the above copyright
+ * notice, this list of conditions and the following disclaimer in the
+ * documentation and/or other materials provided with the distribution;
+ * neither the name of the copyright holders nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "mem/ruby/structures/ALUFreeListArray.hh"
+
+#include "base/intmath.hh"
+#include "mem/ruby/system/RubySystem.hh"
+#include "sim/cur_tick.hh"
+
+#include <stdio.h>
+
+
+namespace gem5
+{
+
+namespace ruby
+{
+
+/*
+* Behavior:
+*   Requests will go through unless one/both of the following are met:
+*       - There have been more than [numALUs] requests in the current cycle
+*       - The same line has been accessed in the past accessLatency cycles
+*/
+
+ALUFreeListArray::ALUFreeListArray(unsigned int numALUs, Cycles accessLatency,
+                                   RubySystem *rs)
+    : m_ruby_system(rs)
+{
+    this->numALUs = numALUs;
+    this->accessLatency = accessLatency;
+}
+
+bool ALUFreeListArray::tryAccess(Addr addr)
```

Comment:
I'm not convinced this logic actually implements the behavior you're going for. I don't understand why things stay in the queue for some amount of time, but you only check for things in the queue at exactly `curTick`.

Could you describe the behavior you're trying to simulate in this function? With this knowledge, I may be able to help you implement what you're going for.

Reply:
The idea is that there are `numALUs` ALU pipelines that are `accessLatency` cycles deep. A new request can enter each pipeline each cycle, which means that each cycle `numALUs` requests can be added (hence why I'm checking for things at the queue in curTick). `accessLatency` is only being used to make sure that a request to a cache line being processed in a pipeline isn't started until the previous request to that cache line makes it all the way through.

Reply:
I'm going to resolve this conversation, but I *still* think the code is not doing what you intend. See my comment below.